vowpalwabbit/learner.h

```diff
@@ -128,6 +135,174 @@ inline void decrement_offset(multi_ex& ec_seq, const size_t increment, const siz
+    else                                                                                                             \
+    {                                                                                                                \
+      std::vector<polylabel> saved;                                                                                  \
+      saved.reserve(example_collection.size());                                                                      \
+      for (size_t i = 0; i < example_collection.size(); i++)                                                         \
+      {                                                                                                              \
+        const auto& example = example_collection[i];                                                                 \
+        saved.push_back(std::move(example->l));                                                                      \
+        example->l.FIELD_NAME = std::move(labels[i]);                                                                \
+      }                                                                                                              \
+      learn_fd.predict_f(learn_fd.data, *learn_fd.base, static_cast<void*>(&example_collection));                    \
+      for (size_t i = 0; i < example_collection.size(); i++)                                                         \
+      {                                                                                                              \
+        const auto& example = example_collection[i];                                                                 \
+        labels[i] = std::move(example->l.FIELD_NAME);                                                                \
+        example->l = std::move(saved[i]);                                                                            \
+      }                                                                                                              \
+    }                                                                                                                \
+  }
+
+#define DEFINE_TYPED_PREDICT_AND_LEARN_FUNCS(TYPE, FIELD_NAME) \
+  TYPED_EXAMPLE_LEARN_FN(TYPE, FIELD_NAME)                     \
+  TYPED_EXAMPLE_PREDICT_FN(TYPE, FIELD_NAME)                   \
+  TYPED_MULTI_EXAMPLE_LEARN_FN(TYPE, FIELD_NAME)               \
+  TYPED_MULTI_EXAMPLE_PREDICT_FN(TYPE, FIELD_NAME)
+
+#define TYPED_EXAMPLE_LEARN_CASE(TYPE, FIELD_NAME)                                                                   \
+  case label_type_t::FIELD_NAME:                                                                                     \
+  {                                                                                                                  \
+    TYPE saved = std::move(ec.l.FIELD_NAME);                                                                         \
```

Comment:
>TYPE saved = std::move(ec.l.FIELD_NAME) [](start = 4, length = 39)

what's the expectation with this code?

for example if label is a vector, will it pass an empty vector to learn?

Reply:
We will be passing a reference to the object that was in `FIELD_NAME` of the polylabel. It must  be stored in a temporary object because we must be able to restore state and the underlying reduction may overwrite the union

Reply:
If the reduction does not use the label but calls base.learn(), would this work correctly?

---
In reply to: [390379556](https://github.com/VowpalWabbit/vowpal_wabbit/pull/2323#discussion_r390379556) [](ancestors = 390379556)

Reply:
In that case the reduction must forward the label to its own learn call

Reply:
So once a reduction starts using label, all reductions invoked by it have to use label and so on down the chain?

---
In reply to: [390393147](https://github.com/VowpalWabbit/vowpal_wabbit/pull/2323#discussion_r390393147) [](ancestors = 390393147)

Reply:
Once we are done with this overall feature, every learn function must have a label as input and must call learn with a label.

As of right now, if a reduction opts in to the new signature they must accept a label as argument and call base.learn with a label. If they have not opted in yet they do not need to.

> So once a reduction starts using label, all reductions invoked by it have to use label and so on down the chain?
No, each reduction is independent on if they use or old style and it doesn't enforce things down the chain at the moment.

Reply:
I was trying to think about the logical downstream implications of modifying ec.l.  Since not all reductions know about label.   I am wondering if its better to not modify ec.l.

---
In reply to: [390379556](https://github.com/VowpalWabbit/vowpal_wabbit/pull/2323#discussion_r390379556) [](ancestors = 390379556)

Reply:
So the point of these macros (and this overall project) is to enable us to not worry about downstream implications. From practice I have not yet found a reduction that this does not work for (I am through about 60% of them)

If you do not modify ec.l then there is no way for a new style reduction to call an old style reduction. Because old style is expecting ec.l to be a certain value where as new style does not touch ec.l at all by design.

This saving and restoring will **go away once the project is done,** I just need it to be able to make reasonable progress in the feature branch.